### PR TITLE
Android support

### PR DESCRIPTION
**Pull request by mzdk100**
This is an initial commit that only implements the Runtime on Android.

### CODE REVIEW

File: winio-ui-android/src/runtime.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        let env = vm.attach_current_thread()?;
+        let activity = unsafe { JObject::from_raw(app.activity_as_ptr() as _) };
+        f(env, activity)
+    }
+
+    fn enter<T, F: FnOnce() -> T>(&self, f: F) -> T {
+        RUNTIME.set(self, f)
+    }
+
+    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
```
Comment (Berrysoft) [inline winio-ui-android/src/runtime.rs:55]: Please refer to `winio-pollable` and `winio-ui-winui` for a correct implementation of `block_on`.

* We don't alloc for `future`. We just call `spawn_unchecked`.
* We need to poll the driver fd and the app event together. If it's not possible, we poll the driver fd in another thread. Your current implementation is enough for GUI to run correctly, but the IO operations will not complete.

File: mobile-examples/android/run.bat
Comment (Berrysoft) [inline mobile-examples/android/run.bat:1]: Just a suggestion: maybe better to use Makefile?

File: winio-ui-android/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+jni.workspace = true
+image = { workspace = true, default-features = false }
+futures-util = "0.3.31"
+
+winio-primitive = { workspace = true }
+winio-callback = { workspace = true }
+winio-handle = {workspace = true}
+winio-pollable = { workspace = true }
+
+compio = { workspace = true, features = [] }
```
Comment (Berrysoft) [inline winio-ui-android/Cargo.toml:24]:
```suggestion
compio = { workspace = true }
```

File: winio-ui-android/src/ui/label.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    pub fn halign(&self) -> HAlign {
+        todo!()
+    }
+
+    pub fn set_halign(&mut self, align: HAlign) {
+        // 通过JNI设置TextView的文本对齐方式
+        RUNTIME.with(|rt| {
+            let view = self.tv.clone();
+            rt.vm_exec(|mut env, _act| {
```
Comment (Berrysoft) [inline winio-ui-android/src/ui/label.rs:47]: Just comment: seems too painful to implement.

File: winio/Cargo.toml [hunk 4]
```diff
@@ -31,7 +31,10 @@ winio-ui-winui = { workspace = true, optional = true }
 [target.'cfg(target_os = "macos")'.dependencies]
 winio-ui-app-kit = { workspace = true }
 
-[target.'cfg(not(any(windows, target_os = "macos")))'.dependencies]
+[target.'cfg(target_os="android")'.dependencies]
+winio-ui-android = { workspace = true, optional = true }
```
Comment (Berrysoft) [inline winio/Cargo.toml:35]: It doesn't have to be an optional dependency.

**Review by Berrysoft — CHANGES_REQUESTED** [review]
* Please rebase the PR.
* Please add `.idea` to `.gitignore`.

I have left some comments. The runtime needs to be modified to make sure `compio` works well.

**Comment by mzdk100** [comment]
Okay, I need some time to study.

**Comment by Berrysoft** [comment]
I understand about the thread checking. But anyway the future is running on the same thread as `block_on`. If it's valid to call `vm_exec` inside the future, it should be valid to call `vm_exec` inside `block_on`.

**Comment by mzdk100** [comment]
> If it's valid to call vm_exec inside the future, it should be valid to call vm_exec inside block_on
That's right, the future and block_on belong to the same thread. But now this thread is not the expected thread for android_activity::AndroidApp.

**Comment by Berrysoft** [comment]
If it's not the correct thread, how do you call `vm_exec` in the correct thread?

**Comment by mzdk100** [comment]
I should have investigated thoroughly, it's not a block_on issue.

File: Cargo.toml [hunk 5]
```diff
@@ -20,7 +20,7 @@ members = [
 resolver = "2"
 
 [workspace.package]
-edition = "2021"
+edition = "2024"
```
Comment (Berrysoft) [inline Cargo.toml:23]: Better not. It's not recommended to use edition = 2024 in libs, because it limits the supported rustc version.

**Review by Berrysoft — COMMENTED** [review]
Please remove the changes not relative to Android support.

File: Cargo.toml (hunk 5 above)
Comment (mzdk100) [reply Cargo.toml:22]: It's okay, edition2024 will be compatible with older versions of Rustc, and many of the new features it brings are very practical.

Comment (Berrysoft) [reply Cargo.toml:22]: Please explain with more details that why do you need edition 2024.

File: winio-ui-android/src/runtime.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    fn enter<T, F: FnOnce() -> T>(&self, f: F) -> T {
+        self.inner.enter(|| RUNTIME.set(self, f))
+    }
+
+    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
+        self.enter(|| {
+            self.inner.block_on(future, |timeout| {
+                sleep(timeout.unwrap_or_else(|| Duration::from_millis(10)));
+            })
```
Comment (Berrysoft) [inline winio-ui-android/src/runtime.rs:32]: Let me help you:
```suggestion
            let mut result = None;
            unsafe {
                self.runtime.spawn_unchecked(async {
                    result = Some(future.await);
                    // exit the application here, if possible
                })
            }
            .detach();
            // start the application here and wait till it exits
            result.unwrap()
```
Then, in the application code, maybe `onStart`:
```rust
compio::runtime::spawn_blocking(move || {
    loop {
        let timeout = resume_foreground(move || {
            RUNTIME.with(|runtime| {
                runtime.runtime.poll_with(Some(Duration::ZERO));
                let remaining_tasks = runtime.run();
                if remaining_tasks {
                    Some(Duration::ZERO)
                } else {
                    runtime.runtime.current_timeout()
                }
            })
        });
        let Some(timeout) = timeout else {
            break;
        };
        // call epoll_wait with timeout value
    }
})
.detach();
```
The method `resume_foreground` executes the passed fn on the UI thread. The winui3 backend gives an example:
```rust
fn resume_foreground<T: Send + 'static>(
    dispatcher: &DispatcherQueue,
    mut f: impl (FnMut() -> T) + Send + 'static,
) -> Option<T> {
    let (tx, rx) = oneshot::channel();
    let mut tx = Some(tx);
    dispatcher
        .TryEnqueue(&DispatcherQueueHandler::new(move || {
            if let Some(tx) = tx.take() {
                tx.send(f()).ok();
            }
            Ok(())
        }))
        .unwrap();
    rx.recv().ok()
}
```

**Review by Berrysoft — CHANGES_REQUESTED** [review]

**Comment by Berrysoft** [comment]
* It's OK to leave monitor utilities as unimplemented.
* But we do need the canvas components. Unfortunately I'm not good at Android developments.

**Comment by mzdk100** [comment]
Can you help me research it? Because I have a visual disability, it's difficult for me to debug images

**Comment by mzdk100** [comment]
The commit of 183a8e7 is incorrect, so I will restore it (2f049ce). The reason is that we cannot directly hold the Android main thread loop, but let the runtime and Android main thread loop run in different threads, otherwise it will be difficult for us to achieve the required functionality.
Now you can run and test the example without any issues.

**Comment by Berrysoft** [comment]
I have said that the preferred design is to *run* the runtime in the main thread, but *poll* the runtime in a separate thread. https://github.com/compio-rs/winio/commit/183a8e7558c5ef03f5fcd93e459017a76496c561 is incorrect because you didn't replace the comments with valid code, and didn't add other necessary codes to the right places.

> Because I have a visual disability, it's difficult for me to debug images

Well, I have noticed that you have completed `Canvas.java`, right? Just don't leave `carvas.rs` with a lot of `todo!()`s, please.

**Comment by Berrysoft** [comment]
> I believe they offer a more developer-friendly approach, safer design, and eventually, we will upgrade to the 2024 edition anyway.

Oh no. Please don't include irrelevant changes in this PR.

If you're not good at English, let's discuss in Chinese:

这个 PR 是和 Android 支持相关的，请不要更改其它代码。支持 edition 2024 的编译器还不够多，我不希望切换过去。即使切换过去，也不应该在这个 PR 里。请解释你具体需要什么 edition 2024 的功能，如果没有就请换回 edition 2021.

> I can understand your design pattern, but it's difficult to implement on Android because we don't have full control over the entire application lifecycle.

我并没有要求控制整个 lifecycle，也不需要控制事件循环。如果你可以让一部分函数在 UI 线程运行，就能让 `runtime.run()` 在 UI 线程运行。

**Comment by mzdk100** [comment]
好的，让我们使用中文来讨论。
我理解您的顾虑，2024版本可能会导致一些编译器不可用，或者遇到一些兼容性问题，并且我们的PR应该按照“问题相关性”来开展，不过我确实很想使用2024版本中的功能，并且当前代码也已经大量使用了，我可以保证就当前项目而言，编译毫无问题，您如果可以先帮我迁移到2024，我会很乐意删除其中的不相关代码。
关于安卓的线程UI，他确实无法实现您的需要，你看我们的入口点是main()函数，我们在其中使用了一个app.run()的方法启动了一个runtime的阻塞循环，假设这个main被放在安卓原生Activity.onCreate事件中调用，那很抱歉，UI将永远卡柱，并且我们没有可用的"pool"方法来解决他，因为onCreate函数没有返回，正确的安卓Activity使用方法是onCreate事件中只能运行业务逻辑，不能启动事件循环，如果您有相关安卓开发经验你会了解到。

**Comment by Berrysoft** [comment]
请您不要再用 LLM 了。我看到代码注释中有中文，因此认为您会更擅长中文。

> 不过我确实很想使用2024版本中的功能，并且当前代码也已经大量使用了

您始终在回避问题：到底想用 edition 2024 的什么功能？

> 假设这个main被放在安卓原生Activity.onCreate事件中调用

为什么要在 `onCreate` 里面调用 `main` 呢？我对 NDK 的 NativeActivity 了解不多，但是程序的入口点应该是可以用 Rust 管理的。

**Comment by mzdk100** [comment]
问题在于驱动main函数的时机，如果过早，那我们的App.run()占用了当前线程，安卓框架将没有机会创建Looper，如果等安卓onCreate后执行main，这会导致安卓onCreate方法无法立刻返回，所以我采用了另一种设计方案，在onCreate中启动单独的线程来驱动我们的main函数，因此这将解决这个问题。
代价就是系统开启了两个并行的事件循环（我们的runtime也拥有完整的生命周期），关于您的疑问，rust确实可以在main中来驱动NativeActivity，但这只是使用NativeActivity，对其他类型的Activity不适用，并且我试验之后他会遇到各种线程相关问题，并且我们无法遵循最佳实践来开展工作，例如编译出的java代码无法正常工作，总之您如果感兴趣您可以迁出这个分支从我第一个提交来运行查看，目前这种方案确实也不是最佳方案，但他确实是容易实施且稳定可用的。
关于2024版本，我认为让代码与rust最新编译器同步是一个很好的做法，他能尽可能减少错误且可以消除任何安全漏洞，2024对const，指针等都有深入的优化，更好的安全性保证，同时`Let chains`是一个强大的工具，他可以显著减少大括号的内嵌，让我们更容易编写代码且具有更好的可读性。

**Comment by mzdk100** [comment]
是的，我使用了一些Let chains，但应该不止于此，没关系，您说的也是正确的

**Comment by Berrysoft** [comment]
> 但他并不需要额外的poll来轮询安卓的looper，没有这个必要

compio 的功能需要 poll。

> 如果2024不符合您的要求，您可以再合并后回退相关代码，这样对此分支保留是可以的。

那我也可以选择不合并这个 PR。贡献代码的是您，审核代码的是我。头一回听说 contributor 给 reviewer 提修改意见的。

**Comment by Berrysoft** [comment]
> 是的，我使用了一些Let chains，但应该不止于此，没关系，您说的也是正确的

我看到一开始用的并不是 edition 2024，是后来才升级的。如果没有明确的原因，或者只是为了使用 let chains，那还是改回去吧。

**Comment by mzdk100** [comment]
okok，我们需要友好交流，其实您的顾虑可能真不存在呢，你可以重新审视一下我们的问题，也许您应该使用CI验证一下。

**Comment by Berrysoft** [comment]
那我再重复一遍：请说明为什么要用 edition 2024？我只看到了如下理由：
* 想要用最新的版本。对不起这是个规模比较大的项目，升级 edition 我有自己的安排，请不要将您的喜好加在这个项目上。
* 想要用 let chains。那还是不要用了，这个问题又不是很难解决。
* 想要用别的功能，请说明。

现在这个 PR 因为 edition 的更改搞得一团糟，无法正常 review。我只能先指出这个主要问题。

**Comment by Berrysoft** [comment]
@mzdk100 对了不知道你有没有发现其实现在 winio 支持了 android，你开的 issue 已经关上了
